### PR TITLE
Initial version of Nanoleaf Aurora control

### PR DESCRIPTION
**Pull request by peternewman**
An initial version if you fancy giving it a first pass.

Still todo/future features:
- [ ] Enable streaming mode
- [ ] Populate panel IDs automatically
- [ ] Populate streaming port automatically
- [ ] Starting offset (each unit is only a max of 30 channels) and/or RDM patching (per controller or per panel)
- [ ] Pass panel vector as part of SendDMX call, so we only need one Node
- [ ] Work out if we need to bother with Rx/binding to the socket

### CODE REVIEW

File: plugins/nanoleaf/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+Nanoleaf Plugin
+============
+
+This plugin creates a device per IP, each with one output port. Each port
```
Comment (nomis52) [inline plugins/nanoleaf/README.md]: Could you add a link to the product.

File: plugins/nanoleaf/NanoleafDevice.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+  if (!m_node->Start()) {
+    delete m_node;
+    m_node = NULL;
+    return false;
+  }
+
+  uint16_t ip_port;
+  if (!StringToInt(m_preferences->GetValue(IPPortKey()), &ip_port)) {
+    ip_port = 60221;
```
Comment (nomis52) [inline plugins/nanoleaf/NanoleafDevice.cpp]: I'd move this to a const kDefaultPort.

File: plugins/nanoleaf/NanoleafDevice.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+string NanoleafDevice::PanelsKey() const {
+  return m_controller.ToString() + "-panels";
+}
+
+
+void NanoleafDevice::SetDefaults() {
+  // Set device options
+  m_preferences->SetDefaultValue(PanelsKey(), StringValidator(), "");
+  m_preferences->SetDefaultValue(IPPortKey(), UIntValidator(1, 65535),
+                                 60221);
```
Comment (nomis52) [inline plugins/nanoleaf/NanoleafDevice.cpp]: use the const.

File: plugins/nanoleaf/NanoleafNode.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ private:
+    bool m_running;
+    ola::io::SelectServerInterface *m_ss;
+    std::vector<uint8_t> m_panels;
+    ola::io::IOQueue m_output_queue;
+    ola::io::OutputStream m_output_stream;
+    ola::network::Interface m_interface;
+    std::auto_ptr<ola::network::UDPSocketInterface> m_socket;
+
+    NanoleafNode(const NanoleafNode&);
```
Comment (nomis52) [inline plugins/nanoleaf/NanoleafNode.h]: Don't we have a macro for these?

Comment (peternewman) [reply plugins/nanoleaf/NanoleafNode.h]: I was briefly confused, having just copied from KiNetNode.h, but yes, this must be DISALLOW_COPY_AND_ASSIGN, I'll fix it up.

Comment (peternewman) [reply plugins/nanoleaf/NanoleafNode.h]: Done, along with other ones that had been missed over the years.

File: plugins/nanoleaf/README.md (hunk 1 above)
Comment (peternewman) [reply plugins/nanoleaf/README.md]: Will do, https://nanoleaf.me/en/consumer-led-lighting/products/smarter-series/nanoleaf-light-panels-smarter-kit/ if you fancy buying some :smile:

File: plugins/nanoleaf/NanoleafDevice.cpp (hunk 2 above)
Comment (peternewman) [reply plugins/nanoleaf/NanoleafDevice.cpp]: Done

File: plugins/nanoleaf/NanoleafDevice.cpp (hunk 3 above)
Comment (peternewman) [reply plugins/nanoleaf/NanoleafDevice.cpp]: Done.

**Comment by peternewman** [comment]
This is ready for re-review @nomis52 . I guess it might make sense to merge this in, then I'll work on the other features later.

**Comment by peternewman** [comment]
Merging while failing, due to issues in base code fixed later.

**Pull request merged**